File: legacy/lib/data/database/daos/customer_dao.dart
import 'package:drift/drift.dart';
import '../app_database.dart';
import '../tables.dart';

part 'customer_dao.g.dart';

@DriftAccessor(tables: [Customers])
class CustomerDao extends DatabaseAccessor<AppDatabase> with _$CustomerDaoMixin {
  CustomerDao(super.db);

  Stream<List<CustomerEntity>> watchAll() =>
      (select(customers)..orderBy([(t) => OrderingTerm(expression: t.updatedAt, mode: OrderingMode.desc)]))
          .watch();

  Stream<List<CustomerEntity>> watchTodayFollowUps(DateTime today) {
    final tomorrow = today.add(const Duration(days: 1));
    return (select(customers)
          ..where((t) => t.nextFollowUpAt.isBiggerOrEqualValue(today) & t.nextFollowUpAt.isSmallerThanValue(tomorrow))
          ..orderBy([(t) => OrderingTerm(expression: t.nextFollowUpAt)]))
        .watch();
  }

  Future<List<CustomerEntity>> getAll() =>
      (select(customers)..orderBy([(t) => OrderingTerm(expression: t.updatedAt, mode: OrderingMode.desc)])).get();

  Future<CustomerEntity?> getById(String id) =>
      (select(customers)..where((t) => t.id.equals(id))).getSingleOrNull();

  Stream<CustomerEntity?> watchById(String id) =>
      (select(customers)..where((t) => t.id.equals(id))).watchSingleOrNull();

  Future<String> insertCustomer(CustomersCompanion entry) async {
    final id = entry.id.present ? entry.id.value : _uuid();
    await into(customers).insert(entry.copyWith(id: Value(id)));
    return id;
  }

  Future<void> updateCustomer(String id, CustomersCompanion entry) async {
    await (update(customers)..where((t) => t.id.equals(id))).write(entry);
  }

  Future<void> deleteCustomer(String id) async {
    await (delete(customers)..where((t) => t.id.equals(id))).go();
  }

  String _uuid() {
    return DateTime.now().microsecondsSinceEpoch.toString() +
        (DateTime.now().millisecond).toString().padLeft(3, '0');
  }
}

File: app/src/main/java/com/salesquest/sales_quest/data/dao/CustomerDao.kt
package com.salesquest.sales_quest.data.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import com.salesquest.sales_quest.data.entity.CustomerEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface CustomerDao {

    @Query("SELECT * FROM customers ORDER BY updatedAt DESC")
    fun watchAll(): Flow<List<CustomerEntity>>

    @Query(
        "SELECT * FROM customers WHERE nextFollowUpAt >= :start AND nextFollowUpAt < :end ORDER BY nextFollowUpAt"
    )
    fun watchTodayFollowUps(start: Long, end: Long): Flow<List<CustomerEntity>>

    @Query("SELECT * FROM customers ORDER BY updatedAt DESC")
    suspend fun getAll(): List<CustomerEntity>

    @Query("SELECT * FROM customers WHERE id = :id")
    suspend fun getById(id: String): CustomerEntity?

    @Query("SELECT * FROM customers WHERE id = :id")
    fun watchById(id: String): Flow<CustomerEntity?>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertCustomer(customer: CustomerEntity)

    @Update
    suspend fun updateCustomer(customer: CustomerEntity)

    @Query("DELETE FROM customers WHERE id = :id")
    suspend fun deleteCustomer(id: String)

    @Query("DELETE FROM customers")
    suspend fun clearAll()

    /**
     * 查询当前最大客户编号 (如 #005 → "#005"), 不依赖客户数量
     * 按数值语义排序: 先按字符串长度(位数)降序, 同长度再按字典序,
     * 避免字典序缺陷 ("#999" > "#1000") 导致编号回退/重复
     */
    @Query(
        "SELECT customerNumber FROM customers WHERE customerNumber IS NOT NULL AND customerNumber LIKE '#%' " +
            "ORDER BY LENGTH(customerNumber) DESC, customerNumber DESC LIMIT 1"
    )
    suspend fun getMaxCustomerNumber(): String?
}
